### PR TITLE
feat: record loss for IVF and KMeans

### PR DESCRIPTION
**Pull request by BubbleCal**
Signed-off-by: BubbleCal <bubble-cal@outlook.com>

### CODE REVIEW

File: rust/lance-index/src/vector/ivf/builder.rs [hunk 1]
```diff
@@ -45,9 +49,6 @@ pub struct IvfBuildParams {
 
     pub shuffle_partition_concurrency: usize,
 
-    /// Use residual vectors to build sub-vector.
-    pub use_residual: bool,
```
Comment (BubbleCal) [inline rust/lance-index/src/vector/ivf/builder.rs:49]: this is never used, so remove it

**Comment by changhiskhan** [comment]
You might want this to be controllable from the optimize call so that no one gets a nasty surprise and block for a long time when they're expecting a quick finish

File: rust/lance-index/src/vector/ivf/transform.rs [hunk 2]
```diff
@@ -84,16 +87,20 @@ impl Transformer for PartitionTransformer {
...
 
-        let part_ids = self.compute_partitions(fsl);
+        let (part_ids, loss) =
+            compute_partitions_arrow_array(&self.centroids, fsl, self.distance_type)?;
+        let part_ids = UInt32Array::from(part_ids);
         let field = Field::new(PART_ID_COLUMN, part_ids.data_type().clone(), true);
-        Ok(batch.try_with_column(field, Arc::new(part_ids))?)
+        Ok(batch
+            .try_with_column(field, Arc::new(part_ids))?
+            .add_metadata(LOSS_METADATA_KEY.to_owned(), loss.to_string())?)
```
Comment (westonpace) [inline rust/lance-index/src/vector/ivf/transform.rs:103]: Do we put the loss in both the index file metadata and the protobuf for the index?

File: rust/lance-index/src/vector/kmeans.rs [hunk 3]
```diff
@@ -16,6 +18,7 @@ use lance_linalg::{
 /// Train KMeans model and returns the centroids of each cluster.
 #[allow(clippy::too_many_arguments)]
 pub fn train_kmeans<T: ArrowPrimitiveType>(
+    centroids: Option<Arc<FixedSizeListArray>>,
```
Comment (westonpace) [inline rust/lance-index/src/vector/kmeans.rs:21]: Can we document what this parameter means?  What happens if `centroids` is provided?  I am guessing we use them as the initial state for the kmeans calculation?

File: rust/lance/src/index/vector/ivf/v2.rs [hunk 4]
```diff
@@ -909,27 +934,145 @@ mod tests {
...
+            .optimize_indices(&OptimizeOptions::retrain())
+            .await
+            .unwrap();
+        let stats = dataset.index_statistics("vector_idx").await.unwrap();
+        let stats: serde_json::Value = serde_json::from_str(&stats).unwrap();
+        assert_eq!(stats["num_indices"], 1);
+
+        let ivf_models = get_ivf_models(&dataset).await;
+        let ivf = &ivf_models[0];
+        assert_ne!(original_ivf.centroids, ivf.centroids);
```
Comment (westonpace) [inline rust/lance/src/index/vector/ivf/v2.rs:1073]: Should we check the loss at this point and make sure it went back down?

File: protos/index.proto [hunk 5]
```diff
@@ -66,6 +66,9 @@ message IVF {
 
   // Tensor of centroids. `num_partitions * dimension` of float32s.
   Tensor centroids_tensor = 4;
+
+  // KMeans loss.
+  optional double loss = 5;
```
[the remark this answers is not in the document]
Comment (westonpace) [reply protos/index.proto:71]: Ok, from looking at the code I think this is `the loss that we experienced after running kmeans`.  It is an output from the training, not an input.

**Review by westonpace — COMMENTED** [review]
Some initial questions but it looks pretty good.  Can we also add `loss` to the statistics for the index so that users can see it?

File: python/python/lance/dataset.py [hunk 6]
```diff
@@ -3561,6 +3561,15 @@ def optimize_indices(self, **kwargs):
...
             If None, all indices will be optimized.
+        retrain: bool, default False
+            Whether to retrain the whole index.
+            If true, the index will be retrained based on the current data,
+            `num_indices_to_merge` will be ignored,
+            and all indices will be merged into one.
+
+            This is useful when the data distribution has changed significantly,
+            and we want to retrain the index to improve the search quality.
+            This would be faster than re-create the index from scratch.
```
[the remark this answers is not in the document]
Comment (BubbleCal) [reply python/python/lance/dataset.py:3572]: The new IVF centroids would be trained based on the previous centroids so it's faster than create_index

File: rust/lance-index/src/vector/ivf/transform.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/ivf/transform.rs:103]: Only protobuf,
Here the transformer stores it in the batch metadata temporally

File: rust/lance-index/src/vector/kmeans.rs (hunk 3 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/kmeans.rs:21]: Yes, will doc it

**Review by westonpace — APPROVED** [review]
